added support for XEP-0030 with tests

andresinaka opened the pull request.

chrisballinger commented:
@andresinaka This looks good, however, I wonder if this is duplicate functionality from the XMPPCapabilities module `- (void)fetchCapabilitiesForJID:(XMPPJID *)jid;`

https://github.com/robbiehanson/XMPPFramework/blob/f9fa43e00f9669839f2b7050010d791109ebad4a/Extensions/XEP-0115/XMPPCapabilities.h#L112

Which internally calls `- (void)sendDiscoInfoQueryTo:(XMPPJID *)jid withNode:(NSString *)node ver:(NSString *)ver` https://github.com/robbiehanson/XMPPFramework/blob/f9fa43e00f9669839f2b7050010d791109ebad4a/Extensions/XEP-0115/XMPPCapabilities.m#L850

It looks like according to [XEP-0115] it should be used instead of XEP-0030 to prevent the "disco/version flood", although in this case the stanzas look nearly identical so I'm not sure how it differs unless the hashed capabilities storage is used.

andresinaka commented:
@chrisballinger I totally agree with you! I added the disco#info to this because I needed to disco#items and I couldn't find a way of doing it. So I googled and found that XEP-0030 so I implemented the whole thing but I think the only missing feature is the disco#items. 

Is there an other way to find out the items from a server?? (this is what i needed: ```[<item jid="muc.my-server.com"/>, <item jid="muclight.my-server.com"/>, <item jid="pubsub.my-server.com"/>, <item jid="vjud.my-server.com"/>]```

Maybe I can add this to some other place? What do you think? 

(btw, don't merge this as is because I found a little but, but first let's discuss about info#items)

chrisballinger commented:
Ahh I see, the #items thing is what you really need. It would be kind of convoluted to use XMPPCapabilities for just the #info part within the XEP-0030 module, and #items is definitely not part of XEP-0115 so it doesn't really belong over there either.

I guess just a mention in the docs that if you're just discovering capabilities you might want to use XMPPCapabilities instead because hashed caps are faster.

chrisballinger commented:
Instead of a full module perhaps you could make a category on `XMPPIQ+XEP_0030.h` to handle generation and parsing of #items responses and defer the work to the client implementation.

andresinaka commented:
@chrisballinger thanks! I think the ```XMPPIQ+XEP_0030.h``` makes way more sense than adding the full XEP-0030! I'll work on that when I have some time!

👍 👍 👍 👍 👍 👍 👍

andresinaka commented:
@chrisballinger I'm back to this. So I created this category that has a method to create the iq for the disco items and a method that parses the response to get the items. 

What do you think?

chrisballinger commented on Extensions/XEP-0030/XMPPIQ+XEP_0030.m:
```diff
@@ -0,0 +1,49 @@
+	NSXMLElement *query = [iq elementForName:@"query" xmlns: XMLNS_DISCO_ITEMS];
+	if(query) {
+		return [query elementsForName:@"item"];
+	}
```
what happens if !query? Also violates nonnull

chrisballinger commented on Extensions/XEP-0030/XMPPIQ+XEP_0030.m:
```diff
@@ -0,0 +1,49 @@
+#import "XMPPIQ+XEP_0030.h"
+#import "NSXMLElement+XMPP.h"
+
+#define XMLNS_DISCO_ITEMS  @"http://jabber.org/protocol/disco#items"
```
https://github.com/robbiehanson/XMPPFramework/blob/f9fa43e00f9669839f2b7050010d791109ebad4a/Core/XMPPConstants.h#L15

chrisballinger commented:
Oh interesting... looks like this is already sort-of implemented inside the file transfer module, albeit tied to the internals:

https://github.com/robbiehanson/XMPPFramework/blob/f9fa43e00f9669839f2b7050010d791109ebad4a/Extensions/FileTransfer/XMPPOutgoingFileTransfer.m#L472

https://github.com/robbiehanson/XMPPFramework/blob/f9fa43e00f9669839f2b7050010d791109ebad4a/Extensions/FileTransfer/XMPPOutgoingFileTransfer.m#L1172

I now see how it could be useful as a module because of the PID tracker, but the standalone IQ parser/generators are still a good thing. Hmmmmmm.

andresinaka commented:
@chrisballinger I like the module solution more because of the PID tracker too. Also we would have the whole XEP-0030 implemented. Let me know what you prefer, if you think the module is better I'll reset hard (i know it's not happy but reverting is going to add an ugly commit and I'm pretty sure I'm the only one working on this) to the module work, fix an small bug and PR it again. 

If you think is better to go with the category I'll fix those problems you found and let you know...

chrisballinger commented:
Given more thought the module seems better especially if it will be growing in size and using the PID tracker, but I'd still think keeping the IQ parser/generators separate would still be good and you could keep the tests you already wrote.

andresinaka commented:
Perfect! I'll do that. I like the idea of keeping the parser/generators separate! Thanks!!

chrisballinger commented:
I fixed a few warnings and nullability issues, and discovered some OS X issues as well: https://github.com/robbiehanson/XMPPFramework/commit/991b5dc3e39b47f80eabaadb219c62ef060f90d1

Please check that there's no warnings or test failures for both iOS and the new macOS test projects. Thanks!
